adapters/triplelift_native/triplelift_native.go

```diff
@@ -0,0 +1,195 @@
+	// pre-process the imps
+	for _, imp := range tlRequest.Imp {
+		if err := processImp(&imp); err == nil {
+			validImps = append(validImps, imp)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+	publisher := getPublisher(request)
+	publisherID := effectivePubID(publisher)
+	if _, exists := a.extInfo.PublisherWhitelistMap[publisherID]; !exists {
+		err := fmt.Errorf("Unsupported publisher for triplelift_native")
+		return nil, []error{err}
+	}
+	if len(validImps) == 0 {
+		err := fmt.Errorf("No valid impressions for triplelift")
+		errs = append(errs, err)
+		return nil, errs
+	}
+	tlRequest.Imp = validImps
+	reqJSON, err := json.Marshal(tlRequest)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	ad := a.endpoint
+	reqs = append(reqs, &adapters.RequestData{
+		Method:  "POST",
+		Uri:     ad,
+		Body:    reqJSON,
+		Headers: headers})
+	return reqs, errs
+}
+
+func getPublisher(request *openrtb.BidRequest) *openrtb.Publisher {
+	if request.App != nil {
+		return request.App.Publisher
+	}
+	return request.Site.Publisher
+}
+
+func getBidCount(bidResponse openrtb.BidResponse) int {
+	c := 0
+	for _, sb := range bidResponse.SeatBid {
+		c = c + len(sb.Bid)
+	}
+	return c
+}
+
+func (a *TripleliftNativeAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+	var errs []error
+	count := getBidCount(bidResp)
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(count)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+			var bidExt TripleliftRespExt
+			bidType := getBidType(bidExt)
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: bidType,
+			})
+		}
+	}
+	return bidResponse, errs
+}
+
+func NewTripleliftNativeBidder(client *http.Client, endpoint string, extraInfo string) *TripleliftNativeAdapter {
+	var extInfo TripleliftNativeExtInfo
+
+	if len(extraInfo) == 0 {
+		extraInfo = "{\"publisher_whitelist\":[]}"
+	}
+	if err := json.Unmarshal([]byte(extraInfo), &extInfo); err != nil {
+		panic("Invalid TripleLife Native extra adapter info: " + err.Error())
```

Comment:
Why is panicking the best option here? Other adapters make their `NewFooBidder()` function throw errors like:
```
    if err := json.Unmarshal([]byte(extraInfo), &extInfo); err != nil {
 -      panic("Invalid TripleLife Native extra adapter info: " + err.Error())
 +      glog.Errorf("Invalid TripleLife Native extra adapter info: " + err.Error())
 +      return &adapters.MisconfiguredBidder{
 +          Name:  "TripleliftNativeAdapter",
 +          Error: errors.New("TripleliftNativeAdapter .... blah, blah, blah"),
 +      }
    }
```
Isn't it better?

Reply:
The panic is from my guidance. I didn't realize we already had a pattern for this sort of thing and the panic seemed appropriate to me for a server configuration issue. This solution looks better though. If it's not a hard ask, could you please follow the guidance from @guscarreon?

Reply:
ok, I followed your advice here.